From cfbcf3f73ac85a64faeeb0d7effd0e90320c4fef Mon Sep 17 00:00:00 2001
From: Karthikeyan K <karthikk@habana.ai>
Date: Mon, 24 May 2021 22:07:34 +0300
Subject: [PATCH] [SW-42526] ras: add fw version support in preboot

In direct command code 104, offset 106-108 is updated with
the preboot version.

106:major version
107:minor version
108:patch version

Change-Id: I40a5f71adddba7a1fcef49526f364d889dc73e35
Signed-off-by: Karthikeyan K <karthikk@habana.ai>
---

diff --git a/Makefile b/Makefile
index 37a4a52..ca936b8 100644
--- a/Makefile
+++ b/Makefile
@@ -91,6 +91,9 @@
 
 $(AUTO_GENERATED_FILE): localversion-habanalabs $(PREBOOT_SRCS_C)
 	@( printf '#define PREBOOT_VER "%s"\n' '$(shell cut -d- -f2- localversion-habanalabs)' ) > $@
+	@( printf '#define PREBOOT_MAJOR_VER %d\n' '$(shell cut -d'-' -f4 localversion-habanalabs | cut -d'.' -f1)' ) >> $@
+	@( printf '#define PREBOOT_MINOR_VER %d\n' '$(shell cut -d'-' -f4 localversion-habanalabs | cut -d'.' -f2)' ) >> $@
+	@( printf '#define PREBOOT_PATCH_VER %d\n' '$(shell cut -d'-' -f4 localversion-habanalabs | cut -d'.' -f3)' ) >> $@
 	@LANG=en_us date +'#define PREBOOT_DATE "%b %d %Y"' >> $@
 	@date +'#define PREBOOT_TIME "%T"' >> $@
 
diff --git a/common/module_i2c_slave.c b/common/module_i2c_slave.c
index 6058f7f..adafc93 100644
--- a/common/module_i2c_slave.c
+++ b/common/module_i2c_slave.c
@@ -268,6 +268,31 @@
 	return ret;
 }
 
+u8 i2cslave_version_cmd(u8 cmd)
+{
+	u8 ret;
+
+	switch (cmd) {
+	case HL_VERSIONS_CMD_CODE:
+		ret = HL_VERSIONS_CMD_LEN;
+		break;
+	case HL_VERSIONS_FW_MAJOR_VER_VAL:
+		ret = PREBOOT_MAJOR_VER;
+		break;
+	case HL_VERSIONS_FW_MINOR_VER_VAL:
+		ret = PREBOOT_MINOR_VER;
+		break;
+	case HL_VERSIONS_FW_ADDITIONAL_VER_VAL:
+		ret = PREBOOT_PATCH_VER;
+		break;
+	default:
+		ret = 0;
+		break;
+	}
+
+	return ret;
+}
+
 #define ERR_STS_REG_MASK_LOW	0xFF
 #define ERR_STS_REG_MASK_HI	0xFF00
 u8 i2cslave_error_status_cmd(u8 cmd)
diff --git a/common/module_i2c_slave.h b/common/module_i2c_slave.h
index 79153f7..bbd7ed0 100644
--- a/common/module_i2c_slave.h
+++ b/common/module_i2c_slave.h
@@ -18,6 +18,7 @@
 
 u8 i2cslave_pcie_vendor_id_cmd(u8 cmd);
 u8 i2cslave_asic_sn_cmd(u8 cmd);
+u8 i2cslave_version_cmd(u8 cmd);
 u8 i2cslave_error_status_cmd(u8 cmd);
 u8 i2cslave_handle_read_req(u8 cmd, u32 read_seq);
 u32 i2cslave_handle_write_req(u8 cmd, u8 **data, u8 cnt);
diff --git a/preboot/module_i2c_slave.c b/preboot/module_i2c_slave.c
index edd0b7a..37f1a20 100644
--- a/preboot/module_i2c_slave.c
+++ b/preboot/module_i2c_slave.c
@@ -65,6 +65,12 @@
 	case HL_ASIC_SERIAL_NUMBER_0 ... HL_ASIC_SERIAL_NUMBER_17:
 		ret = i2cslave_asic_sn_cmd(cmd);
 		break;
+	case HL_VERSIONS_CMD_CODE:
+	case HL_VERSIONS_FW_MAJOR_VER_VAL:
+	case HL_VERSIONS_FW_MINOR_VER_VAL:
+	case HL_VERSIONS_FW_ADDITIONAL_VER_VAL:
+		ret = i2cslave_version_cmd(cmd);
+		break;
 	case HL_ERROR_STATUS_CODE:
 		ret = HL_ERROR_STATUS_CMD_LEN;
 		break;
